### PR TITLE
Opt: Delete names and decos of dead instructions

### PR DESCRIPTION
**Pull request by greg-lunarg**
Add named_or_decorated_ids cache to speed up name and deco kills.
GroupDecorate not supported and will prohibit transform.

### CODE REVIEW

File: source/opt/dead_branch_elim_pass.cpp [hunk 1]
```diff
@@ -147,8 +147,35 @@ void DeadBranchElimPass::AddBranchConditional(uint32_t condId,
...
+  // kill if no other operands.
+  if (named_or_decorated_ids_.find(id) == named_or_decorated_ids_.end())
+    return;
+  analysis::UseList* uses = def_use_mgr_->GetUses(id);
+  if (uses == nullptr)
+    return;
+  std::list<ir::Instruction*> killList;
+  for (auto u : *uses) {
+    const SpvOp op = u.inst->opcode();
+    if (op != SpvOpName && !IsDecorate(op))
```
Comment (antiagainst) [inline source/opt/dead_branch_elim_pass.cpp]: Isn't this just
```cpp
if (op == SpvOpname || IsDecorate(op))
  killList.push_back(u.inst);
```
?

File: source/opt/dead_branch_elim_pass.h [hunk 2]
```diff
@@ -97,6 +97,21 @@ class DeadBranchElimPass : public Pass {
   // Return true if |labelId| has any non-phi references
   bool HasNonPhiRef(uint32_t labelId);
 
+  // Return true if all uses of varId are only through supported reference
```
Comment (antiagainst) [inline source/opt/dead_branch_elim_pass.h]: This comment is not relevant to the method.

File: source/opt/local_access_chain_convert_pass.h [hunk 3]
```diff
@@ -72,6 +72,24 @@ class LocalAccessChainConvertPass : public Pass {
   // variables.
   bool IsTargetVar(uint32_t varId);
 
+  // Return true if all uses of varId are only through supported reference
```
Comment (antiagainst) [inline source/opt/local_access_chain_convert_pass.h]: Irrelevant comment.

File: source/opt/local_single_block_elim_pass.h [hunk 4]
```diff
@@ -81,6 +81,24 @@ class LocalSingleBlockLoadStoreElimPass : public Pass {
   // Add stores using |ptr_id| to |insts|
   void AddStores(uint32_t ptr_id, std::queue<ir::Instruction*>* insts);
 
+  // Return true if all uses of varId are only through supported reference
```
Comment (antiagainst) [inline source/opt/local_single_block_elim_pass.h]: Same here

File: source/opt/local_single_block_elim_pass.cpp [hunk 5]
```diff
@@ -165,6 +166,48 @@ void LocalSingleBlockLoadStoreElimPass::AddStores(
...
+  // kill if no other operands.
+  if (named_or_decorated_ids_.find(id) == named_or_decorated_ids_.end())
+    return;
+  analysis::UseList* uses = def_use_mgr_->GetUses(id);
+  if (uses == nullptr)
+    return;
+  std::list<ir::Instruction*> killList;
+  for (auto u : *uses) {
+    const SpvOp op = u.inst->opcode();
+    if (op != SpvOpName && !IsDecorate(op))
```
Comment (antiagainst) [inline source/opt/local_single_block_elim_pass.cpp]: Same here

File: source/opt/local_single_store_elim_pass.cpp [hunk 6]
```diff
@@ -345,6 +347,48 @@ void LocalSingleStoreElimPass::AddStores(
...
+  // kill if no other operands.
+  if (named_or_decorated_ids_.find(id) == named_or_decorated_ids_.end())
+    return;
+  analysis::UseList* uses = def_use_mgr_->GetUses(id);
+  if (uses == nullptr)
+    return;
+  std::list<ir::Instruction*> killList;
+  for (auto u : *uses) {
+    const SpvOp op = u.inst->opcode();
+    if (op != SpvOpName && !IsDecorate(op))
```
Comment (antiagainst) [inline source/opt/local_single_store_elim_pass.cpp]: Same here

File: source/opt/local_single_store_elim_pass.h [hunk 7]
```diff
@@ -119,6 +119,24 @@ class LocalSingleStoreElimPass : public Pass {
   // Add stores using |ptr_id| to |insts|
   void AddStores(uint32_t ptr_id, std::queue<ir::Instruction*>* insts);
 
+  // Return true if all uses of varId are only through supported reference
```
Comment (antiagainst) [inline source/opt/local_single_store_elim_pass.h]: Same here

File: source/opt/local_access_chain_convert_pass.cpp [hunk 8]
```diff
@@ -127,6 +143,32 @@ void LocalAccessChainConvertPass::ReplaceAndDeleteLoad(
...
+  // kill if no other operands.
+  if (named_or_decorated_ids_.find(id) == named_or_decorated_ids_.end())
+    return;
+  analysis::UseList* uses = def_use_mgr_->GetUses(id);
+  if (uses == nullptr)
+    return;
+  std::list<ir::Instruction*> killList;
+  for (auto u : *uses) {
+    const SpvOp op = u.inst->opcode();
+    if (op != SpvOpName && !IsDecorate(op))
```
Comment (antiagainst) [inline source/opt/local_access_chain_convert_pass.cpp]: Same here

**Comment by greg-lunarg** [comment]
In general, sorry for the repetition. This hasn't been a huge problem in the past, but now it is getting to the point of needing change. Getting rid of this will be a top priority.

I am thinking of creating an OptPass which derives from Pass which would contain all the common code. Then all my new Passes would derive off that.

File: source/opt/dead_branch_elim_pass.cpp (hunk 1 above)
Comment (greg-lunarg) [inline source/opt/dead_branch_elim_pass.cpp]: Yes. I will change to that. Everywhere :)

File: source/opt/dead_branch_elim_pass.h (hunk 2 above)
Comment (greg-lunarg) [inline source/opt/dead_branch_elim_pass.h]: Gah! How did that happen? Will fix. Everywhere :(

**Comment by greg-lunarg** [comment]
I think all issues have been addressed. Let me know if you need anything else.

**Comment by antiagainst** [comment]
Rebased, squashed and merged as https://github.com/KhronosGroup/SPIRV-Tools/commit/7954740d542786acd071310a8978ffb3e2042b7c

**Comment by antiagainst** [comment]
Thanks @greg-lunarg!

Yep, the duplication is growing. I agree with you that some kind of common "context" should be introduced to solve this problem.

**Pull request closed without merging**